amethyst_error/src/lib.rs

```diff
@@ -0,0 +1,449 @@
+    ///     Err(Error::from_string("failing"))
+    /// }
+    ///
+    /// fn other_function() -> Result<(), Error> {
+    ///     Ok(failing_function().with_context(|_| Error::from_string("other"))?)
+    /// }
+    ///
+    /// let e = other_function().expect_err("no error");
+    ///
+    /// let messages = e.causes().map(|e| e.to_string()).collect::<Vec<_>>();
+    /// assert_eq!(vec!["other", "failing"], messages);
+    /// ```
+    pub fn causes(&self) -> Causes<'_> {
+        Causes {
+            current: Some(self),
+        }
+    }
+
+    /// Access the internal `std::error::Error` as a trait.
+    ///
+    /// This can be useful for integrating with systems that operate on `std::error::Error`.
+    ///
+    /// **Warning:** This erases most diagnostics in favor of returning only the top error.
+    /// `std::error::Error` is expanded further.
+    pub fn as_error(&self) -> &(dyn error::Error + 'static) {
+        &self.inner.error
+    }
+}
+
+/// Blanket implementation.
+///
+/// Encapsulate errors which are Send + Sync.
+impl<T> From<T> for Error
+where
+    T: 'static + error::Error + Send + Sync,
+{
+    fn from(value: T) -> Error {
+        Error::new(value)
+    }
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt::Display::fmt(&self.inner.error, fmt)
+    }
+}
+
+impl fmt::Debug for Error {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt.debug_struct("Error")
+            .field("inner", &self.inner)
+            .finish()
+    }
+}
+
+/// Extra convenience functions for results based on core errors.
+pub trait ResultExt<T>
+where
+    Self: Sized,
+{
+    /// Provide a context for the result in case it is an error.
+    ///
+    /// The context callback is expected to return a new error, which will replace the given error
+    /// and set the replaced error as its [`source`](Error::source).
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use amethyst_error::{Error, ResultExt};
+    ///
+    /// fn failing_function() -> Result<(), Error> {
+    ///     Err(Error::from_string("failing"))
+    /// }
+    ///
+    /// fn other_function() -> Result<(), Error> {
+    ///     Ok(failing_function().with_context(|_| Error::from_string("other"))?)
+    /// }
+    ///
+    /// let e = other_function().expect_err("no error");
+    ///
+    /// assert_eq!("other", e.to_string());
+    /// assert_eq!("failing", e.source().expect("no source").to_string());
+    /// assert!(e.source().expect("no source").source().is_none());
+    /// ```
+    fn with_context<C, D>(self, chain: C) -> Result<T, Error>
+    where
+        C: FnOnce(&Error) -> D,
+        D: Into<Error>;
+}
+
+impl<T, E> ResultExt<T> for result::Result<T, E>
+where
+    E: Into<Error>,
+{
+    fn with_context<C, D>(self, chain: C) -> Result<T, Error>
+    where
+        C: FnOnce(&Error) -> D,
+        D: Into<Error>,
```

Comment:
`D` should impl `Into<Error>` but `error::Error` instead.

Reply:
Converting into `Error` is to guarantee capturing a backtrace at these composition points. Similarly to how `with_context` returns `Context<E>` in failure.

Reply:
Capturing a backtrace is exactly what I want to avoid in this function.